### PR TITLE
feat: Added Exponential Probability Distribution

### PR DESCRIPTION
**Pull request by HarshilShah1804**
#### Description of Change

Added Exponential distribution in probability.
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: probability/exponential_dist.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/**
+ * @file
+ * @brief [Exponential Distribution] (https://en.wikipedia.org/wiki/Exponential_distribution)
+ * 
+ * The exponential distribution is used to model
+ * events occuring between a Poisson process like radioactive decay.
+ * 
+ * P(x,⋋) = ⋋e^(-⋋x)
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:8]: use latex notation instead of the raw latin alphabet

File: probability/exponential_dist.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ * 
+ * Summary of variables used:
+ * ⋋: rate parameter
+ */
+#include<cmath>
+#include<cassert>
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:14]: headers should be documented

File: probability/exponential_dist.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+#include<iostream>
+
+/** Find the expected value of the exponential distribution
+ * \param [in] lambda
+ * \returns \f$\mu=1/lambda\f$
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:19]: for consistency please use @'s instead of \'s

File: probability/exponential_dist.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    std::cout << "Output : " << exponential_std(lambda) << std::endl;
+    assert(exponential_std(lambda)==std);
+    std::cout << "TEST PASSED" << std::endl;
+    std::cout<<std::endl;
+}
+
+int main(){
+    test(); //Self-implemented test
+    return 0;
+}
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:90]:
```suggestion
}

```

File: probability/exponential_dist.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    std::cout << "Lambda : " << lambda << std::endl;
+    std::cout << "Expected Output : " << std <<std::endl;
+    std::cout << "Output : " << exponential_std(lambda) << std::endl;
+    assert(exponential_std(lambda)==std);
+    std::cout << "TEST PASSED" << std::endl;
+    std::cout<<std::endl;
+}
+
+int main(){
+    test(); //Self-implemented test
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:88]:
```suggestion
    test(); //Self test implementation
```

File: probability/exponential_dist.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    std::cout << "Standard Deviation" << std::endl;
+    std::cout << "Lambda : " << lambda << std::endl;
+    std::cout << "Expected Output : " << std <<std::endl;
+    std::cout << "Output : " << exponential_std(lambda) << std::endl;
+    assert(exponential_std(lambda)==std);
+    std::cout << "TEST PASSED" << std::endl;
+    std::cout<<std::endl;
+}
+
+int main(){
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:87]: main should be documented

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: probability/exponential_dist.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
+/**
+ * @file
+ * @brief [Exponential Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
+ * @brief [Exponential Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
+ * 
+ * The exponential distribution is used to model
+ * events occuring between a Poisson process like radioactive decay.
+ * 
+ * \( P(x, \lambda) = \lambda e^{-\lambda x} \)
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:9]:
```suggestion
 * \f[P(x, \lambda) = \lambda e^{-\lambda x}\f]
```

File: probability/exponential_dist.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ * 
+ * Summary of variables used:
+ * \lambda : rate parameter
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:12]:
```suggestion
 * \f$\lambda\f$ : rate parameter
```

File: probability/exponential_dist.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
+/**
+ * @file
+ * @brief [Exponential Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
+ * @brief [Exponential Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:4]:
```suggestion
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
replace \( with \f$ for it to render
\f[ makes a big equation
 while the other is inline 
 examples can be found [here](https://thealgorithms.github.io/C-Plus-Plus/da/d23/eulers__totient__function_8cpp.html)

File: probability/exponential_dist.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+ * @returns \f[\sigma = \frac{1}{\lambda}\f]
+ */
+double exponential_std(double lambda){
+    if (lambda<=0){
+        std::cout << "Error: Lambda must be greater than 0." << std::endl;
+        assert(lambda>0);
+    }
+    return 1/lambda;
+}
+
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:53]:
```suggestion

/**
 * @brief Self-test implementation
 * @returns void
 */
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: probability/exponential_dist.cpp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+ * The exponential distribution is used to model
+ * events occuring between a Poisson process like radioactive decay.
+ * 
+ * \f[P(x, \lambda) = \lambda e^{-\lambda x}\f]
+ * 
+ * Summary of variables used:
+ * \f$\lambda\f$ : rate parameter
+ */
+
+#include<cmath> // For power function
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:14]:
```suggestion
#include<cmath> // For std::pow
```

File: probability/exponential_dist.cpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+#include<cassert> // For asserting the test cases
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:15]:
```suggestion
#include<cassert> // For assert
```

File: probability/exponential_dist.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+static void test(){
+    double lambda = 2;
+    double expected = 0.5;
+    double var = 0.25;
+    double std = 0.5;
+
+    //Test 1
+    std::cout << "Expected Value" << std::endl;
+    std::cout << "Lambda : " << lambda << std::endl;
+    std::cout << "Expected Output : " << expected <<std::endl;
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:67]: these outputs are a bit unnecessary although not detrimental. you could replace them with more test cases

**Review by realstealthninja — CHANGES_REQUESTED** [review]
please lint your file with clang-tidy / clang-format

File: probability/exponential_dist.cpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    double lambda_2 = 2;
+    double expected_2 = 0.5;
+    double var_2 = 0.25;
+    double std_2 = 0.5;
+
+    double lambda_3 = 3;
+    double expected_3 = 0.333333;
+    double var_3 = 0.111111;
+    double std_3 = 0.333333;
+
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:74]: add some tests with negative numbers and zero

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: probability/exponential_dist.cpp (hunk 14 above)
Comment (HarshilShah1804) [reply probability/exponential_dist.cpp:74]: The value of lambda is always greater than 0 that's why I haven't added them. It is in assertion in the functions.

Comment (realstealthninja) [reply probability/exponential_dist.cpp:74]: your algorithm should throw a runtime error in such case, and you should account for this by asserting that the error is indeed thrown. in a short example it could be implemented as follows
```cpp
int my_algorithm(double my_val) {
   if (my_val < 0) {
      throw std::invalid_argument("my_val cant be negative");
   }
}


...

static void tests() {
   try {
      my_algorithm(-3.14);
   } except(std::invalid_argument& err) {
      assert(err.what() == "my_val cant be negative");
   }
}



```

Comment (HarshilShah1804) [reply probability/exponential_dist.cpp:73]: @realstealthninja  I have added tests for 0 and negative values

**Comment by HarshilShah1804** [comment]
@realstealthninja All changes are made. Please review

File: probability/exponential_dist.cpp [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+ * Summary of variables used:
+ * \f$\lambda\f$ : rate parameter
+ */
+
+#include <cassert>    // For assert
+#include <cmath>      // For std::pow
+#include <iostream>   // For I/O operation
+#include <stdexcept>  // For std::invalid_argument
+#include <string>     // For std::string
+
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:20]: these functions should be inside the probability namespace except for tests

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by HarshilShah1804** [comment]
@realstealthninja I have modified the code according to the suggestions. Please review and tell me if anything else is required. Thank you

File: probability/exponential_dist.cpp [hunk 16]
```diff
@@ -0,0 +1,151 @@
+/**
+ * @file
+ * @brief [Exponential
+ * Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:4]: this should be a single line

File: probability/exponential_dist.cpp [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * @namespace probability
+ * @brief Probability algorithms
+ */
+namespace probability {
+/**
+ * @namespace exponential_dist
+ * @brief Functions for the [Exponential
+ * Distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
```
Comment (realstealthninja) [inline probability/exponential_dist.cpp:29]: should be single line

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Review by realstealthninja — APPROVED** [review]
Lgtm, thank you for your contribution. Happy hacktoberfest ♥️
